src/Monolog/Formatter/SyslogFormatter.php

```diff
@@ -0,0 +1,68 @@
+<?php declare(strict_types=1);
+
+/*
+ * This file is part of the Monolog package.
+ *
+ * (c) Jordi Boggiano <j.boggiano@seld.be>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Monolog\Formatter;
+
+use Monolog\Logger;
+
+/**
+ * Serializes a log message according to RFC 5424
+ *
+ * @author Dalibor Karlović <dalibor.karlovic@sigwin.hr>
+ */
+class SyslogFormatter extends LineFormatter
+{
+    private const SYSLOG_FACILITY_USER = 1;
+    private const LEVELS = [
+        Logger::EMERGENCY => 0,
+        Logger::ALERT => 1,
+        Logger::CRITICAL => 2,
+        Logger::ERROR => 3,
+        Logger::WARNING => 4,
+        Logger::NOTICE => 5,
+        Logger::INFO => 6,
+        Logger::DEBUG => 7,
+    ];
+    private const FORMAT = '<%extra.priority%>1 %datetime% %extra.hostname% %extra.app-name% %extra.procid% %channel% %extra.structured-data% %message%'."\n";
+    private const NILVALUE = '-';
+
+    public function __construct(?string $applicationName = null)
+    {
+        parent::__construct(self::FORMAT, 'Y-m-d\TH:i:s.uP', true);
+
+        $this->extra = [
+            'app-name' => $applicationName ?? self::NILVALUE,
+            'hostname' => (string) gethostname(),
+            'procid' => (int) getmypid(),
+        ];
+    }
+
+    public function format(array $record): string
+    {
+        $record['extra'] = $this->formatExtra($record);
+
+        return parent::format($record);
+    }
+
+    private static function calculatePriority(int $level): int
+    {
+        return (self::SYSLOG_FACILITY_USER * 8) + self::LEVELS[$level];
+    }
+
+    private function formatExtra(array $record): array
+    {
+        $extra = $this->extra;
+        $extra['priority'] = self::calculatePriority($record['level']);
+        $extra['structured-data'] = self::NILVALUE;
+        
+        return $extra;
```

Comment:
Is this getting rid of the original record's extra data on purpose? I'd be better to array_merge it than replace it entirely IMO.

Really I guess this should use a nested LineFormatter to format the record, and use the formatted record as %message% when formatting to syslog format.

Reply:
The target format is very specific about the details, no arbitrary extras can be added or it will no longer be Syslog compatible. Still need to format/add the contextual data (param just before message).

> I guess this should use a nested LineFormatter to format the record, and use the formatted record as %message% when formatting to syslog format.

Not sure what you mean here, it does use the LineFormatter, no?

Reply:
I mean when you use a SyslogHandler it will format a log record with extra/context data using a LineFormatter, resulting in something like `"message {"contextKey": "value"} {"extrakey": "value"}"` in `$record['formatted']`. This whole string then gets passed to syslog in https://github.com/Seldaek/monolog/blob/fd4380d6fc37626e2f799f29d91195040137eba9/src/Monolog/Handler/SyslogHandler.php#L65

While with your formatter here, everything gets lost except for the message. If you created a LineFormatter internally to format the record, and then decorate it with a syslog format, you should be able to keep the same level of functionality.

Reply:
Syslog natively supports structured data, this is still pending to get implemented here
https://datatracker.ietf.org/doc/html/rfc5424#section-6.3

Reply:
Ok, I mean if you want to implement it using that it sounds ok to me too, although the ASCII restriction seems a bit outdated.. I don't really mind how the data makes it in, but IMO context/extra should not disappear.